packages/dds/merge-tree/src/mergeTreeNodes.ts

```diff
@@ -174,8 +174,55 @@ export interface ISegment extends IMergeNodeCommon, Partial<IRemovalInfo> {
 	 */
 	propertyManager?: PropertiesManager;
 	/**
-	 * Local seq at which this segment was inserted. If this is defined, `seq` will be UnassignedSequenceNumber.
+	 * Local seq at which this segment was inserted.
+	 * This is defined if an only if the insertion of the segment is pending ack, i.e. `seq` is UnassignedSequenceNumber.
 	 * Once the segment is acked, this field is cleared.
+	 * Semantically, `localSeq`s provide an ordering on in-flight merge-tree operations:
+	 * for operations stamped with localSeqs `a` and `b`, `a < b` if and only if `a` was submitted before `b`.
+	 *
+	 * @remarks - This field is roughly analogous to the `clientSequenceNumber` field on ops, but it's accessible to merge-tree
+	 * at op submission time rather than only at ack time. This enables more natural state tracking for in-flight ops.
+	 *
+	 * It's useful to stamp ops with such an incrementing counter because it enables reasoning about which segments existed from
+	 * the perspective of the local client at a given point in 'un-acked' time, which is necessary to support the reconnect flow.
+	 *
+	 * For example, imagine a client with the following segments undergoes the reconnect flow:
+	 * ```js
+	 * [
+	 *     { seq: 0, text: "1234" },
+	 *     { seq: 5, text: "56" },
+	 *     { localSeq: 3, seq: UnassignedSequenceNumber, text: "A" },
+	 *     { localSeq: 2, seq: UnassignedSequenceNumber, text: "B" },
+	 *     { localSeq: 1, seq: UnassignedSequenceNumber, text: "C" },
+	 * ]
+	 * ```
+	 * This client submitted the "C" first, then the "B" and finally the "A".
+	 *
+	 * Since it's in the reconnect flow, it needs to resubmit those 3 ops to the server based on the initial state.
+	 * The ops it submits will be:
+	 * 1. \{ pos: 6, text: "C" \}
+	 * 2. \{ pos: 6, text: "B" \}
+	 * 3. \{ pos: 6, text: "A" \}
+	 *
+	 * since when submitting the first op, remote clients don't know that this client is about to submit the "A" and "B".
```

Comment:
"C" is the first op.

> Since the localSeq of "C" is less than that of "B" and "A", shouldn't the local client know about "C" at the time "B" is submitted?

This is correct, but it doesn't affect the insertion position of "B", since "C" is positioned after "B".

Anything I can do to make the statements here more clear?

Reply:
Oh, does it depend on the actual positions of the inserted text before resubmit as well? I didn't realize that, but if it inserts to try to get back to the state of "123456ABC" then it makes sense.

Reply:
Maybe elaborating on the scenario a bit more in the setup might help. Here we're considering a scenario where a client wants to go from the state "123456" to the state "123456ABC". They might take several different paths to go between those two states. The first path would involve first inserting "C", then "B", then "A" whereas the second path would be "A", then "B", then "C".

In either case, if their 3 ops get nacked and undergo the resubmission flow, they need to submit equivalent ops in the same order, and the goal is to submit ops which create as desirable of a merge outcome as possible. You can see that localSeq on segments impacts the position of insertion that happens upon resubmission.

...the example is a *little* silly, since I didn't want to get into details of the resubmission flow having to rebase the original op: for this reason, the resubmitted ops look exactly like the originally generated ops. If you want to consider that kind of scenario instead, imagine instead that when originally submitting A/B/C, the client only saw the state "12",  then in the meantime while waiting for ack of their op, a remote client inserted the "3456". It doesn't really affect anything with what localSeq is meant for, which is why I didn't want to elaborate on it much. But there might be simpler schemes for doing this kind of thing in absence of the fact that we need to support remote editing, so it's one of those logical leaps that might not be obvious to someone trying to understand for the first time.

Reply:
Ohh ok, that helps a lot! I think a little more elaboration in the setup would be good, but knowing that, the rest makes sense now.

Reply:
took a stab at tweaking the prose. I also moved the example onto CollaborationWindow.localSeq and pointed the relevant ISegment fields at it, which I think makes more sense pedagogically.